File: AC-tually Cold Enough/gemini2.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <queue>

using namespace std;

void solve() {
    int n;
    long long d;
    int k;
    if (!(cin >> n >> d >> k)) return;
    
    vector<long long> h(n);
    for (int i = 0; i < n; ++i) {
        cin >> h[i];
    }
    
    vector<vector<int>> adj(n);
    for (int i = 0; i < n - 1; ++i) {
        int u, v;
        cin >> u >> v;
        adj[u].push_back(v);
        adj[v].push_back(u);
    }
    
    if (k > n) {
        cout << -1 << "\n";
        return;
    }
    
    vector<long long> req(n);
    req[0] = h[0];
    queue<pair<int, int>> q; 
    q.push({0, -1});
    vector<long long> depth(n, 0);
    
    while (!q.empty()) {
        auto [u, p] = q.front();
        q.pop();
        for (int v : adj[u]) {
            if (v != p) {
                depth[v] = depth[u] + 1;
                req[v] = max(req[u], h[v] + depth[v] * d);
                q.push({v, u});
            }
        }
    }
    
    sort(req.begin(), req.end());
    cout << req[k - 1] << "\n";
}

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);
    solve();
    return 0;
}

File: AC-tually Cold Enough/gemini2.py
import sys

def solve():
    input_data = sys.stdin.read().split()
    if not input_data:
        return
    
    N = int(input_data[0])
    D = int(input_data[1])
    K = int(input_data[2])
    
    if K > N:
        print("-1")
        return
        
    H = [int(x) for x in input_data[3:3+N]]
    
    idx = 3 + N
    adj = [[] for _ in range(N)]
    for _ in range(N - 1):
        u = int(input_data[idx])
        v = int(input_data[idx+1])
        idx += 2
        adj[u].append(v)
        adj[v].append(u)
        
    req = [0] * N
    req[0] = H[0]
    depth = [0] * N
    
    queue = [0]
    visited = [False] * N
    visited[0] = True
    
    for u in queue:
        for v in adj[u]:
            if not visited[v]:
                visited[v] = True
                depth[v] = depth[u] + 1
                req[v] = max(req[u], H[v] + depth[v] * D)
                queue.append(v)
                
    req.sort()
    print(req[K - 1])

if __name__ == '__main__':
    solve()